File: sdk/typescript/src/rbac.ts
/**
 * Role and permission management — admin API.
 *
 * Wraps:
 *   - `/api/v1/roles` — CRUD + permission attach/detach
 *   - `/api/v1/permissions` — list/create/delete
 *   - `/api/v1/users/{user_id}/roles` and
 *     `/api/v1/users/{user_id}/roles/{rid}` — role assign/revoke
 *
 * See `internal/api/router.go` for the canonical route inventory.
 */

import { httpRequest } from "./http.js";
import { SharkAPIError } from "./errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Role row. */
export interface Role {
  id: string;
  name: string;
  description?: string;
  created_at?: string;
  updated_at?: string;
  [key: string]: unknown;
}

/** Permission row. */
export interface Permission {
  id: string;
  action: string;
  resource: string;
  created_at?: string;
  [key: string]: unknown;
}

/** Input for {@link RbacClient.updateRole}. */
export interface UpdateRoleInput {
  name?: string;
  description?: string;
  [key: string]: unknown;
}

/** Options for {@link RbacClient}. */
export interface RbacClientOptions {
  baseUrl: string;
  adminKey: string;
}

const ROLES = "/api/v1/roles";
const PERMS = "/api/v1/permissions";
const USERS = "/api/v1/users";

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/** Admin client for the global RBAC surface. */
export class RbacClient {
  private readonly _base: string;
  private readonly _key: string;

  constructor(opts: RbacClientOptions) {
    this._base = opts.baseUrl.replace(/\/+$/, "");
    this._key = opts.adminKey;
  }

  private _auth(): Record<string, string> {
    return { Authorization: `Bearer ${this._key}` };
  }

  private async _throw(status: number, text: string): Promise<never> {
    let code = "api_error";
    let message = text.slice(0, 300);
    try {
      const body = JSON.parse(text) as {
        error?: { code?: string; message?: string } | string;
        message?: string;
      };
      if (typeof body.error === "object" && body.error) {
        if (body.error.code) code = body.error.code;
        if (body.error.message) message = body.error.message;
      } else if (typeof body.error === "string") {
        message = body.error;
      } else if (body.message) {
        message = body.message;
      }
    } catch {
      // keep defaults
    }
    throw new SharkAPIError(message, code, status);
  }

  private static _unwrap<T>(body: unknown): T {
    if (
      body &&
      typeof body === "object" &&
      "data" in (body as Record<string, unknown>) &&
      Object.keys(body as Record<string, unknown>).length <= 2
    ) {
      return (body as { data: T }).data;
    }
    return body as T;
  }

  private static _list<T>(body: unknown, ...keys: string[]): T[] {
    if (Array.isArray(body)) return body as T[];
    if (body && typeof body === "object") {
      const obj = body as Record<string, unknown>;
      for (const k of [...keys, "data"]) {
        const v = obj[k];
        if (Array.isArray(v)) return v as T[];
      }
    }
    return [];
  }

  // ------------------------------------------------------------- Roles

  /** GET /api/v1/roles. */
  async listRoles(): Promise<Role[]> {
    const url = `${this._base}${ROLES}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return RbacClient._list<Role>(resp.json(), "roles");
  }

  /** POST /api/v1/roles. */
  async createRole(name: string, description?: string): Promise<Role> {
    const body: Record<string, unknown> = { name };
    if (description !== undefined) body.description = description;
    const url = `${this._base}${ROLES}`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    return RbacClient._unwrap<Role>(resp.json());
  }

  /** GET /api/v1/roles/{id}. */
  async getRole(roleId: string): Promise<Role> {
    const url = `${this._base}${ROLES}/${encodeURIComponent(roleId)}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return RbacClient._unwrap<Role>(resp.json());
  }

  /** PUT /api/v1/roles/{id} (backend exposes update as PUT, not PATCH). */
  async updateRole(roleId: string, fields: UpdateRoleInput): Promise<Role> {
    const url = `${this._base}${ROLES}/${encodeURIComponent(roleId)}`;
    const resp = await httpRequest(url, {
      method: "PUT",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(fields),
    });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return RbacClient._unwrap<Role>(resp.json());
  }

  /** DELETE /api/v1/roles/{id}. */
  async deleteRole(roleId: string): Promise<void> {
    const url = `${this._base}${ROLES}/${encodeURIComponent(roleId)}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204)
      await this._throw(resp.status, resp.text);
  }

  // ----------------------------------------------------- Permissions

  /** GET /api/v1/permissions. */
  async listPermissions(): Promise<Permission[]> {
    const url = `${this._base}${PERMS}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return RbacClient._list<Permission>(resp.json(), "permissions");
  }

  /** POST /api/v1/permissions. */
  async createPermission(action: string, resource: string): Promise<Permission> {
    const url = `${this._base}${PERMS}`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify({ action, resource }),
    });
    if (resp.status !== 200 && resp.status !== 201)
      await this._throw(resp.status, resp.text);
    return RbacClient._unwrap<Permission>(resp.json());
  }

  /** POST /api/v1/roles/{role_id}/permissions. */
  async attachPermission(roleId: string, permissionId: string): Promise<void> {
    const url = `${this._base}${ROLES}/${encodeURIComponent(roleId)}/permissions`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify({ permission_id: permissionId }),
    });
    if (![200, 201, 204].includes(resp.status))
      await this._throw(resp.status, resp.text);
  }

  /** DELETE /api/v1/roles/{role_id}/permissions/{permission_id}. */
  async detachPermission(roleId: string, permissionId: string): Promise<void> {
    const url = `${this._base}${ROLES}/${encodeURIComponent(
      roleId,
    )}/permissions/${encodeURIComponent(permissionId)}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204)
      await this._throw(resp.status, resp.text);
  }

  // --------------------------------------------- User <-> role assignment

  /** POST /api/v1/users/{user_id}/roles. */
  async assignRole(userId: string, roleId: string): Promise<void> {
    const url = `${this._base}${USERS}/${encodeURIComponent(userId)}/roles`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify({ role_id: roleId }),
    });
    if (![200, 201, 204].includes(resp.status))
      await this._throw(resp.status, resp.text);
  }

  /** DELETE /api/v1/users/{user_id}/roles/{role_id}. */
  async revokeRole(userId: string, roleId: string): Promise<void> {
    const url = `${this._base}${USERS}/${encodeURIComponent(
      userId,
    )}/roles/${encodeURIComponent(roleId)}`;
    const resp = await httpRequest(url, { method: "DELETE", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204)
      await this._throw(resp.status, resp.text);
  }

  // --------------------------------------------- Read-only helpers

  /** GET /api/v1/users/{user_id}/roles. */
  async listUserRoles(userId: string): Promise<Role[]> {
    const url = `${this._base}${USERS}/${encodeURIComponent(userId)}/roles`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return RbacClient._list<Role>(resp.json(), "roles");
  }

  /** GET /api/v1/users/{user_id}/permissions. */
  async listUserPermissions(userId: string): Promise<Permission[]> {
    const url = `${this._base}${USERS}/${encodeURIComponent(userId)}/permissions`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return RbacClient._list<Permission>(resp.json(), "permissions");
  }
}

File: sdk/python/shark_auth/rbac.py
"""Role and permission management — admin API.

Wraps:

- ``/api/v1/roles`` — CRUD + permission attach/detach
- ``/api/v1/permissions`` — list/create/delete
- ``/api/v1/users/{user_id}/roles`` and
  ``/api/v1/users/{user_id}/roles/{rid}`` — role assign/revoke

See ``internal/api/router.go`` for the canonical route inventory.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from . import _http
from .proxy_rules import _raise


class RBACClient:
    """Admin client for the global RBAC surface.

    Parameters
    ----------
    base_url:
        Base URL of the SharkAuth server.
    admin_api_key:
        Admin API key (``sk_live_...``).
    session:
        Optional pre-configured :class:`requests.Session`.
    """

    _ROLES = "/api/v1/roles"
    _PERMS = "/api/v1/permissions"
    _USERS = "/api/v1/users"

    def __init__(
        self,
        base_url: str,
        admin_api_key: str,
        *,
        session: Optional[object] = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._token = admin_api_key
        self._session = session or _http.new_session()

    def _auth(self) -> Dict[str, str]:
        return {"Authorization": f"Bearer {self._token}"}

    @staticmethod
    def _unwrap(body: Any) -> Any:
        if isinstance(body, dict) and "data" in body and len(body) <= 2:
            return body["data"]
        return body

    @staticmethod
    def _list(body: Any, *keys: str) -> List[Dict[str, Any]]:
        if isinstance(body, list):
            return body
        if isinstance(body, dict):
            for k in (*keys, "data"):
                v = body.get(k)
                if isinstance(v, list):
                    return v
        return []

    # ------------------------------------------------------------------
    # Roles
    # ------------------------------------------------------------------

    def list_roles(self) -> List[Dict[str, Any]]:
        """GET /api/v1/roles."""
        url = f"{self._base}{self._ROLES}"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            return self._list(resp.json(), "roles")
        _raise(resp)

    def create_role(
        self, name: str, description: Optional[str] = None
    ) -> Dict[str, Any]:
        """POST /api/v1/roles."""
        body: Dict[str, Any] = {"name": name}
        if description is not None:
            body["description"] = description
        url = f"{self._base}{self._ROLES}"
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code in (200, 201):
            return self._unwrap(resp.json())
        _raise(resp)

    def get_role(self, role_id: str) -> Dict[str, Any]:
        """GET /api/v1/roles/{id}."""
        url = f"{self._base}{self._ROLES}/{role_id}"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            return self._unwrap(resp.json())
        _raise(resp)

    def update_role(self, role_id: str, **fields: Any) -> Dict[str, Any]:
        """PUT /api/v1/roles/{id}.

        The backend exposes update as PUT (not PATCH) — see router.go.
        """
        url = f"{self._base}{self._ROLES}/{role_id}"
        resp = _http.request(self._session, "PUT", url, headers=self._auth(), json=fields)
        if resp.status_code == 200:
            return self._unwrap(resp.json())
        _raise(resp)

    def delete_role(self, role_id: str) -> None:
        """DELETE /api/v1/roles/{id}."""
        url = f"{self._base}{self._ROLES}/{role_id}"
        resp = _http.request(self._session, "DELETE", url, headers=self._auth())
        if resp.status_code in (200, 204):
            return None
        _raise(resp)

    # ------------------------------------------------------------------
    # Permissions
    # ------------------------------------------------------------------

    def list_permissions(self) -> List[Dict[str, Any]]:
        """GET /api/v1/permissions."""
        url = f"{self._base}{self._PERMS}"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            return self._list(resp.json(), "permissions")
        _raise(resp)

    def create_permission(self, action: str, resource: str) -> Dict[str, Any]:
        """POST /api/v1/permissions."""
        body = {"action": action, "resource": resource}
        url = f"{self._base}{self._PERMS}"
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code in (200, 201):
            return self._unwrap(resp.json())
        _raise(resp)

    def attach_permission(self, role_id: str, permission_id: str) -> None:
        """POST /api/v1/roles/{role_id}/permissions."""
        url = f"{self._base}{self._ROLES}/{role_id}/permissions"
        resp = _http.request(
            self._session,
            "POST",
            url,
            headers=self._auth(),
            json={"permission_id": permission_id},
        )
        if resp.status_code in (200, 201, 204):
            return None
        _raise(resp)

    def detach_permission(self, role_id: str, permission_id: str) -> None:
        """DELETE /api/v1/roles/{role_id}/permissions/{permission_id}."""
        url = f"{self._base}{self._ROLES}/{role_id}/permissions/{permission_id}"
        resp = _http.request(self._session, "DELETE", url, headers=self._auth())
        if resp.status_code in (200, 204):
            return None
        _raise(resp)

    # ------------------------------------------------------------------
    # User <-> role assignment
    # ------------------------------------------------------------------

    def assign_role(self, user_id: str, role_id: str) -> None:
        """POST /api/v1/users/{user_id}/roles.

        Backend handler: ``handleAssignRole`` (router.go).
        """
        url = f"{self._base}{self._USERS}/{user_id}/roles"
        resp = _http.request(
            self._session,
            "POST",
            url,
            headers=self._auth(),
            json={"role_id": role_id},
        )
        if resp.status_code in (200, 201, 204):
            return None
        _raise(resp)

    def revoke_role(self, user_id: str, role_id: str) -> None:
        """DELETE /api/v1/users/{user_id}/roles/{role_id}.

        Backend handler: ``handleRemoveRole`` (router.go).
        """
        url = f"{self._base}{self._USERS}/{user_id}/roles/{role_id}"
        resp = _http.request(self._session, "DELETE", url, headers=self._auth())
        if resp.status_code in (200, 204):
            return None
        _raise(resp)

    # ------------------------------------------------------------------
    # Read-only helpers
    # ------------------------------------------------------------------

    def list_user_roles(self, user_id: str) -> List[Dict[str, Any]]:
        """GET /api/v1/users/{user_id}/roles."""
        url = f"{self._base}{self._USERS}/{user_id}/roles"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            return self._list(resp.json(), "roles")
        _raise(resp)

    def list_user_permissions(self, user_id: str) -> List[Dict[str, Any]]:
        """GET /api/v1/users/{user_id}/permissions."""
        url = f"{self._base}{self._USERS}/{user_id}/permissions"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            return self._list(resp.json(), "permissions")
        _raise(resp)
